toucan_connectors/common.py

```diff
@@ -385,3 +386,68 @@ def pandas_read_sql(
             raise
 
     return df
+
+
+def format_db_tree(unformatted_db_tree: list):
+    """
+    From [
+    ('database', 'schema', 'table', 'table type', {'column name':'column type'}), ...
+    ]
+    to
+        [
+      {
+        name: "Table 1",
+        type: "table",
+        database: 'Database 1',
+        schema: 'Schema 1',
+        columns: [
+          {
+            name: "ColA",
+            type: "string"
+          }
+        ]
+      },
+      {
+        name: "View 1",
+        type: "view",
+        database: 'Database 1',
+        schema: 'Schema 1',
+        columns: [
+          {
+            name: "ColO",
+            type: "string",
+          }
+        ]
+      },
+    ]
+    """
+    df = pd.DataFrame(unformatted_db_tree)
+    df.columns = ['database', 'schema', 'type', 'name', 'columns']
+    try:  # if columns is a string
+        df['columns'] = df['columns'].apply(json.loads)
+    except TypeError:  # else ignore
+        pass
+    output = []
+
+    for db in df['database'].unique():
+        schemas = df[df['database'] == db]['schema'].unique()
+        for schema in schemas:
+            for type in ('table', 'view'):
+                objects = df[
+                    (df['database'] == db) & (df['schema'] == schema) & (df['type'] == type)
+                ]['name'].unique()
+                for object in objects:
+                    current_object = {
+                        'name': object,
+                        'schema': schema,
+                        'database': db,
+                        'type': type,
+                        'columns': df[
+                            (df['database'] == db)
+                            & (df['schema'] == schema)
+                            & (df['type'] == type)
+                            & (df['name'] == object)
+                        ]['columns'].values[0],
+                    }
+                    output.append(current_object)
```

Comment:
J'ai l'impression que cette quadruple boucle est overkill et qu'on pourrait s'en sortir avec un unique "for" mais il y a peut être quelque chose qui m'échappe ?
Le nombre d’éléments de `unformatted_db_tree` est le même que `output` ou il y a de l'élagage fait entre temps ?

Reply:
En effet je suis allé au plus simple sans considération d'optimisation. 
Même nombre d'éléments entre entrée & sortie. Y'a ptetre moyen de faire un group by avec une aggrégation en liste pour les colonnes.

Reply:
Ok, j'avais pas de considération d'optimisation non plus, je me demande en quoi ce gros block de code diffère de quelque chose comme : 

```python
[ { 'name': e[0], 'schema': e[1], 'database': e[2]  } for e in unformatted_db_tree ]
```

?